juju/juju
# Make storage detachable/attachable

## Description of change

When detaching storage from a unit, we now
leave the storage in the model, and just
clear the storage instance's ownership. The
storage can then be attached to another unit
using the new State.AttachStorage method.
Hook up the attach-storage backend so it calls
State.AttachStorage.

The existing model storage cleanups have been
replaced with one that destroys storage
instances, rather than volumes and filesystems.
The machine storage will be destroyed as a
consequence.

## QA steps

1. juju bootstrap aws
2. juju deploy storagetest -n2
3. juju add-storage storagetest/0 disks=ebs
(wait)
4. juju detach-storage disks/0
5. juju attach-storage storagetest/1 disks/0
(wait - should fail because storagetest/0 and storagetest/1 are in different AZs)
6. juju detach-storage disks/0
7. juju add-machine zone=(same zone as storagetest/0's machine)
8. juju add-unit storagetest --to (machine ID from step 7)
9. juju attach-storage storagetest/2 disks/0
(wait - should attach cleanly)
10. juju destroy-controller -y aws-...

## Documentation changes

When this branch lands into develop, after 2.2 is released, we need to document detach-storage and attach-storage in the devel docs.

## Bug reference

https://bugs.launchpad.net/bugs/1592887
